File: apps/bff/src/project/project.controller.ts
import { Body, Controller, Delete, Get, Param, Patch, Post, Put, Query, } from '@nestjs/common';
import { ProjectService } from './project.service';
import { AuthHeader } from '../common/decorators/auth-header.decorator';
import type {
  CreateProjectInvitationRequest,
  CreateProjectRequest,
  UpdateProjectMemberRoleRequest,
  UpdateProjectRequest,
} from './project.types';

@Controller('api/v1.0')
export class ProjectController {
  constructor(private readonly projectService: ProjectService) {
  }

  @Post('workspaces/:workspaceId/projects')
  async createProject(
    @Param('workspaceId') workspaceId: string,
    @Body() data: CreateProjectRequest,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.createProject(workspaceId, data, authHeader);
  }

  @Get('workspaces/:workspaceId/projects')
  async getProjects(
    @Param('workspaceId') workspaceId: string,
    @Query('page') page = 0,
    @Query('size') size = 5,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.getProjects(
      workspaceId,
      +page,
      +size,
      authHeader,
    );
  }

  @Get('projects/:projectId')
  async getProject(
    @Param('projectId') projectId: string,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.getProject(projectId, authHeader);
  }

  @Put('projects/:projectId')
  async updateProject(
    @Param('projectId') projectId: string,
    @Body() data: UpdateProjectRequest,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.updateProject(projectId, data, authHeader);
  }

  @Delete('projects/:projectId')
  async deleteProject(
    @Param('projectId') projectId: string,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.deleteProject(projectId, authHeader);
  }

  @Get('projects/:projectId/members')
  async getMembers(
    @Param('projectId') projectId: string,
    @Query('page') page = 0,
    @Query('size') size = 5,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.getMembers(projectId, +page, +size, authHeader);
  }

  @Patch('projects/:projectId/members/:userId/role')
  async updateMemberRole(
    @Param('projectId') projectId: string,
    @Param('userId') userId: string,
    @Body() data: UpdateProjectMemberRoleRequest,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.updateMemberRole(
      projectId,
      userId,
      data,
      authHeader,
    );
  }

  @Delete('projects/:projectId/members/me')
  async leaveProject(
    @Param('projectId') projectId: string,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.leaveProject(projectId, authHeader);
  }

  @Delete('projects/:projectId/members/:userId')
  async removeMember(
    @Param('projectId') projectId: string,
    @Param('userId') userId: string,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.removeMember(projectId, userId, authHeader);
  }

  @Post('projects/:projectId/invitations')
  async createInvitation(
    @Param('projectId') projectId: string,
    @Body() data: CreateProjectInvitationRequest,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.createInvitation(projectId, data, authHeader);
  }

  @Get('projects/:projectId/invitations')
  async getInvitations(
    @Param('projectId') projectId: string,
    @Query('page') page = 0,
    @Query('size') size = 10,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.getInvitations(
      projectId,
      +page,
      +size,
      authHeader,
    );
  }

  @Get('users/me/invitations/projects')
  async getMyInvitations(
    @Query('page') page = 0,
    @Query('size') size = 10,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.getMyInvitations(+page, +size, authHeader);
  }

  @Patch('projects/invitations/:invitationId/accept')
  async acceptInvitation(
    @Param('invitationId') invitationId: string,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.acceptInvitation(invitationId, authHeader);
  }

  @Patch('projects/invitations/:invitationId/reject')
  async rejectInvitation(
    @Param('invitationId') invitationId: string,
    @AuthHeader() authHeader: string,
  ) {
    return this.projectService.rejectInvitation(invitationId, authHeader);
  }
}

File: apps/backend/api/src/main/java/com/schemafy/api/project/controller/ProjectController.java
package com.schemafy.api.project.controller;

import jakarta.validation.Valid;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.PositiveOrZero;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.Authentication;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import com.schemafy.api.common.constant.ApiPath;
import com.schemafy.api.common.type.PageResponse;
import com.schemafy.api.project.controller.dto.request.CreateProjectRequest;
import com.schemafy.api.project.controller.dto.request.UpdateProjectMemberRoleRequest;
import com.schemafy.api.project.controller.dto.request.UpdateProjectRequest;
import com.schemafy.api.project.controller.dto.request.UpdateProjectShareLinkRequest;
import com.schemafy.api.project.controller.dto.response.ProjectMemberResponse;
import com.schemafy.api.project.controller.dto.response.ProjectResponse;
import com.schemafy.api.project.controller.dto.response.ProjectShareLinkResponse;
import com.schemafy.api.project.controller.dto.response.ProjectSummaryResponse;
import com.schemafy.api.project.orchestrator.ProjectMemberOrchestrator;
import com.schemafy.core.project.application.port.in.CreateProjectCommand;
import com.schemafy.core.project.application.port.in.CreateProjectUseCase;
import com.schemafy.core.project.application.port.in.DeleteProjectCommand;
import com.schemafy.core.project.application.port.in.DeleteProjectUseCase;
import com.schemafy.core.project.application.port.in.GetMySharedProjectsQuery;
import com.schemafy.core.project.application.port.in.GetMySharedProjectsUseCase;
import com.schemafy.core.project.application.port.in.GetProjectMembersQuery;
import com.schemafy.core.project.application.port.in.GetProjectQuery;
import com.schemafy.core.project.application.port.in.GetProjectShareLinkQuery;
import com.schemafy.core.project.application.port.in.GetProjectShareLinkUseCase;
import com.schemafy.core.project.application.port.in.GetProjectUseCase;
import com.schemafy.core.project.application.port.in.GetProjectsQuery;
import com.schemafy.core.project.application.port.in.GetProjectsUseCase;
import com.schemafy.core.project.application.port.in.LeaveProjectCommand;
import com.schemafy.core.project.application.port.in.LeaveProjectUseCase;
import com.schemafy.core.project.application.port.in.RemoveProjectMemberCommand;
import com.schemafy.core.project.application.port.in.RemoveProjectMemberUseCase;
import com.schemafy.core.project.application.port.in.UpdateProjectCommand;
import com.schemafy.core.project.application.port.in.UpdateProjectMemberRoleCommand;
import com.schemafy.core.project.application.port.in.UpdateProjectShareLinkCommand;
import com.schemafy.core.project.application.port.in.UpdateProjectShareLinkUseCase;
import com.schemafy.core.project.application.port.in.UpdateProjectUseCase;

import lombok.RequiredArgsConstructor;
import reactor.core.publisher.Mono;

@RestController
@Validated
@RequestMapping(ApiPath.API)
@RequiredArgsConstructor
public class ProjectController {

  private final CreateProjectUseCase createProjectUseCase;
  private final GetProjectsUseCase getProjectsUseCase;
  private final GetMySharedProjectsUseCase getMySharedProjectsUseCase;
  private final GetProjectUseCase getProjectUseCase;
  private final UpdateProjectUseCase updateProjectUseCase;
  private final DeleteProjectUseCase deleteProjectUseCase;
  private final RemoveProjectMemberUseCase removeProjectMemberUseCase;
  private final LeaveProjectUseCase leaveProjectUseCase;
  private final GetProjectShareLinkUseCase getProjectShareLinkUseCase;
  private final UpdateProjectShareLinkUseCase updateProjectShareLinkUseCase;
  private final ProjectMemberOrchestrator projectMemberOrchestrator;

  @Value("${app.base-url:http://localhost:8080}")
  private String baseUrl;

  @PostMapping("/workspaces/{workspaceId}/projects")
  @ResponseStatus(HttpStatus.CREATED)
  public Mono<ProjectResponse> createProject(
      @PathVariable String workspaceId,
      @Valid @RequestBody CreateProjectRequest request,
      Authentication authentication) {
    String userId = authentication.getName();
    return createProjectUseCase.createProject(new CreateProjectCommand(
        workspaceId,
        request.dbVendorId(),
        request.name(),
        request.description(),
        userId))
        .map(ProjectResponse::from);
  }

  @GetMapping("/workspaces/{workspaceId}/projects")
  public Mono<PageResponse<ProjectSummaryResponse>> getProjects(
      @PathVariable String workspaceId,
      @RequestParam(defaultValue = "0") @PositiveOrZero int page,
      @RequestParam(defaultValue = "5") @Positive @Max(100) int size,
      Authentication authentication) {
    String userId = authentication.getName();
    return getProjectsUseCase.getProjects(new GetProjectsQuery(
        workspaceId,
        userId,
        page,
        size))
        .map(result -> PageResponse.of(
            result.content().stream().map(ProjectSummaryResponse::from).toList(),
            result.page(),
            result.size(),
            result.totalElements()));
  }

  @GetMapping("/projects/{projectId}")
  public Mono<ProjectResponse> getProject(
      @PathVariable String projectId,
      Authentication authentication) {
    String userId = authentication.getName();
    return getProjectUseCase.getProject(new GetProjectQuery(projectId, userId))
        .map(ProjectResponse::from);
  }

  @GetMapping("/projects/shared/me")
  public Mono<PageResponse<ProjectSummaryResponse>> getMySharedProjects(
      @RequestParam(defaultValue = "0") @PositiveOrZero int page,
      @RequestParam(defaultValue = "5") @Positive @Max(100) int size,
      Authentication authentication) {
    String userId = authentication.getName();
    return getMySharedProjectsUseCase.getMySharedProjects(
        new GetMySharedProjectsQuery(userId, page, size))
        .map(result -> PageResponse.of(
            result.content().stream()
                .map(ProjectSummaryResponse::from).toList(),
            result.page(),
            result.size(),
            result.totalElements()));
  }

  @PutMapping("/projects/{projectId}")
  public Mono<ProjectResponse> updateProject(
      @PathVariable String projectId,
      @Valid @RequestBody UpdateProjectRequest request,
      Authentication authentication) {
    String userId = authentication.getName();
    return updateProjectUseCase.updateProject(new UpdateProjectCommand(
        projectId,
        request.name(),
        request.description(),
        userId))
        .map(ProjectResponse::from);
  }

  @DeleteMapping("/projects/{projectId}")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public Mono<Void> deleteProject(@PathVariable String projectId,
      Authentication authentication) {
    String userId = authentication.getName();
    return deleteProjectUseCase.deleteProject(new DeleteProjectCommand(
        projectId,
        userId));
  }

  @GetMapping("/projects/{projectId}/members")
  public Mono<PageResponse<ProjectMemberResponse>> getMembers(
      @PathVariable String projectId,
      @RequestParam(defaultValue = "0") @PositiveOrZero int page,
      @RequestParam(defaultValue = "5") @Positive @Max(100) int size,
      Authentication authentication) {
    String userId = authentication.getName();
    return projectMemberOrchestrator.getMembers(new GetProjectMembersQuery(
        projectId,
        userId,
        page,
        size))
        .map(result -> result.map(ProjectMemberResponse::from));
  }

  @PatchMapping("/projects/{projectId}/members/{userId}/role")
  public Mono<ProjectMemberResponse> updateMemberRole(
      @PathVariable String projectId,
      @PathVariable String userId,
      @Valid @RequestBody UpdateProjectMemberRoleRequest request,
      Authentication authentication) {
    String requesterId = authentication.getName();
    return projectMemberOrchestrator.updateMemberRole(
        new UpdateProjectMemberRoleCommand(
            projectId,
            userId,
            request.role(),
            requesterId))
        .map(ProjectMemberResponse::from);
  }

  @DeleteMapping("/projects/{projectId}/members/{userId}")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public Mono<Void> removeMember(@PathVariable String projectId, @PathVariable String userId,
      Authentication authentication) {
    String requester = authentication.getName();
    return removeProjectMemberUseCase.removeProjectMember(
        new RemoveProjectMemberCommand(projectId, userId, requester));
  }

  @DeleteMapping("/projects/{projectId}/members/me")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public Mono<Void> leaveProject(
      @PathVariable String projectId, Authentication authentication) {
    String userId = authentication.getName();
    return leaveProjectUseCase.leaveProject(new LeaveProjectCommand(
        projectId,
        userId));
  }

  @GetMapping("/projects/{projectId}/share-link")
  public Mono<ProjectShareLinkResponse> getProjectShareLink(
      @PathVariable String version,
      @PathVariable String projectId,
      Authentication authentication) {
    String userId = authentication.getName();
    return getProjectShareLinkUseCase.getProjectShareLink(
        new GetProjectShareLinkQuery(projectId, userId))
        .map(link -> ProjectShareLinkResponse.of(link, baseUrl, version))
        .defaultIfEmpty(ProjectShareLinkResponse.inactive());
  }

  @PatchMapping("/projects/{projectId}/share-link")
  public Mono<ProjectShareLinkResponse> updateProjectShareLink(
      @PathVariable String version,
      @PathVariable String projectId,
      @Valid @RequestBody UpdateProjectShareLinkRequest request,
      Authentication authentication) {
    String userId = authentication.getName();
    return updateProjectShareLinkUseCase.updateProjectShareLink(
        new UpdateProjectShareLinkCommand(
            projectId,
            request.isActive(),
            userId))
        .map(link -> ProjectShareLinkResponse.of(link, baseUrl, version))
        .defaultIfEmpty(ProjectShareLinkResponse.inactive());
  }

}
